internal/app/machined/pkg/controllers/block/internal/volumes/locate.go

```diff
@@ -0,0 +1,159 @@
+
+	"github.com/siderolabs/gen/value"
+	"github.com/siderolabs/gen/xerrors"
+	"github.com/siderolabs/go-blockdevice/v2/partitioning"
+	"go.uber.org/zap"
+
+	"github.com/siderolabs/talos/pkg/machinery/cel/celenv"
+	"github.com/siderolabs/talos/pkg/machinery/resources/block"
+)
+
+// LocateAndProvision locates and provisions a volume.
+//
+// It verifies that the volume is either already created, and ready to be used,
+// or provisions it.
+//
+//nolint:gocyclo,cyclop
+func LocateAndProvision(logger *zap.Logger, volumeContext ManagerContext) error {
+	volumeType := volumeContext.Cfg.TypedSpec().Type
+
+	if volumeType == block.VolumeTypeTmpfs {
+		// tmpfs volumes are always ready
+		volumeContext.Status.Phase = block.VolumePhaseReady
+
+		return nil
+	}
+
+	// below for partition/disk volumes:
+	if value.IsZero(volumeContext.Cfg.TypedSpec().Locator) {
+		return fmt.Errorf("volume locator is not set")
+	}
+
+	// attempt to discover the volume
+	for _, dv := range volumeContext.DiscoveredVolumes {
+		matches, err := volumeContext.Cfg.TypedSpec().Locator.Match.EvalBool(celenv.VolumeLocator(), map[string]any{
+			"volume": dv,
+		})
+		if err != nil {
+			return fmt.Errorf("error evaluating volume locator: %w", err)
+		}
+
+		if matches {
+			volumeContext.Status.Phase = block.VolumePhaseLocated
+			volumeContext.Status.Location = dv.DevPath
+			volumeContext.Status.PartitionIndex = int(dv.PartitionIndex)
+			volumeContext.Status.ParentLocation = dv.ParentDevPath
+
+			volumeContext.Status.UUID = dv.Uuid
+			volumeContext.Status.PartitionUUID = dv.PartitionUuid
+			volumeContext.Status.Size = dv.Size
+
+			return nil
+		}
+	}
+
+	if !volumeContext.DevicesReady {
+		// volume wasn't located, but devices are not ready yet
+		volumeContext.Status.Phase = block.VolumePhaseWaiting
+
+		return nil
+	}
+
+	// if we got here, the volume is missing, so it needs to be provisioned
+	if value.IsZero(volumeContext.Cfg.TypedSpec().Provisioning) {
+		// the volume can't be provisioned, because the provisioning instructions are missing
+		volumeContext.Status.Phase = block.VolumePhaseMissing
+
+		return nil
+	}
+
+	if !volumeContext.PreviousWaveProvisioned {
+		// previous wave is not provisioned yet
+		volumeContext.Status.Phase = block.VolumePhaseWaiting
+
+		return nil
+	}
+
+	// locate the disk(s) for the volume
+	var matchedDisks []string
+
+	for _, diskCtx := range volumeContext.Disks {
+		if diskCtx.Disk.Readonly {
+			// skip readonly disks, they can't be provisioned either way
+			continue
+		}
+
+		matches, err := volumeContext.Cfg.TypedSpec().Provisioning.DiskSelector.Match.EvalBool(celenv.DiskLocator(), map[string]any{
+			"disk":        diskCtx.Disk,
+			"system_disk": diskCtx.SystemDisk,
+		})
+		if err != nil {
+			return fmt.Errorf("error evaluating disk locator: %w", err)
+		}
+
+		if matches {
+			matchedDisks = append(matchedDisks, diskCtx.Disk.DevPath)
+		}
+	}
+
+	if len(matchedDisks) == 0 {
+		return fmt.Errorf("no disks matched for volume")
```

Comment:
I wonder if we should log that some readonly disk got matched, but we skipped it.

Reply:
readonly disks can't be used for provisioning, but in the future we can allow having volumes on them (readonly)

Reply:
I'm not sure how useful that is in real life. Read-only disks are e.g. Talos squashfs, CD-ROM? What is the real usecase for _provisioning_ volumes on such devices?